resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/MultiServerFeignTarget.scala

```diff
@@ -54,14 +59,31 @@ private[kubernetes] class MultiServerFeignTarget[T : ClassTag](
     requestTemplate.request()
   }
 
+  override def url(): String = threadLocalShuffledServers.get.head
+
   override def continueOrPropagate(e: RetryableException): Unit = {
-    threadLocalShuffledServers.set(threadLocalShuffledServers.get.drop(1))
-    if (threadLocalShuffledServers.get.isEmpty) {
-      throw e
+    threadLocalCurrentAttempt.set(threadLocalCurrentAttempt.get + 1)
+    val currentAttempt = threadLocalCurrentAttempt.get
+    if (threadLocalCurrentAttempt.get < maxRetriesPerServer) {
+      logWarning(s"Attempt $currentAttempt of $maxRetriesPerServer failed for" +
+        s" server ${url()}. Retrying request...", e)
+      Thread.sleep(delayBetweenRetriesMillis)
```

Comment:
Or rather - everything in Feign is synchronous so there would be no other realistic way to add the delay except for here (unless of course we wanted the retry loop in the caller - but trying to abstract that away with this in the first place)